engine/timer.cpp

```diff
@@ -25,26 +25,25 @@ void Timer::HandleTimerCallbacks()
     const int timeWaitedUs = static_cast<int>(timeWaited.count());
     for (auto& timerRequest : mActiveTimers)
     {
-        timerRequest->mTimeoutUs -= timeWaitedUs;
-        if (timerRequest->mTimeoutUs <= 0)
+        timerRequest.mTimeoutUs -= timeWaitedUs;
+        if (timerRequest.mTimeoutUs <= 0)
         {
-            timerRequest->mCallback();
+            timerRequest.mCallback();
         }
     }
 
     mActiveTimers.erase(std::remove_if(mActiveTimers.begin(), mActiveTimers.end(), [](const auto& timerRequest) {
-        return timerRequest->mTimeoutUs <= 0;
+        return timerRequest.mTimeoutUs <= 0;
     }), mActiveTimers.end());
 }
 
 void Timer::AddTimer(int gameObjectId, int timeoutMs, std::function<void()> callback)
 {
     // Interface uses ms but internally keeps track using us. This is just in case the
-    // game's framerate is uncapped. It is possible that {timeWaited} is less then 1ms
-    // and comes out to 0 then the timer will not proceed.
+    // game's framerate is uncapped. If using ms, it is possible that {timeWaited} is less then 1ms
+    // and comes out to 0. This will cause the timer to not proceed.
     const int timeoutUs = timeoutMs * MS_TO_US;
-    auto timerRequest = std::make_shared<TimerRequest>(gameObjectId, timeoutUs, callback);
-    mActiveTimers.push_back(timerRequest);
+    mActiveTimers.emplace_back(gameObjectId, timeoutUs, callback);
```

Comment:
nice!! much better. 

Any time you use "reference semantics" also known as "pointer semantics" there are a few things that can go wrong

1. performance is way worse, because you are fragmenting memory and can't keep it in the cache
2. You have to deal with the whole "is it valid" nullptr stuff, which can make your code gross
3. You have to explicitly determine ownership, which is it's own can of worms.


Wherever possible, prefer "value semantics" over reference semantics. It makes the code more comprehensible and easier to deal with overall.